submit external tx endpoint in API

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#627 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have extended network and wallet layer to accommodate this new endpoint later
- [x] I have added not supported error to http-bridge network impl
- [x] I have reused getMessage from binary to decode `(Tx, [TxWitness])` from binary blob and then `postTx`
- [x] I have added network tests of `decodeExternalTx` for Jormungandr
- [x] I have added endpoint to API
- [x] I have updated swagger.yaml
- [x] I have added integration tests

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
extend wallet and network layer to accommodate externally signed tx

paweljakubas committed:
use getMessage to decode binary blob and reuse postTx

paweljakubas committed:
add positive property test for decodeExternalTx jormungandr network layer

paweljakubas committed:
using runGetOrFail in impl and add exception throwing property to network spec of Jormungandr

nix patch

paweljakubas committed:
add endpoint to API

paweljakubas committed:
add missing tests to TypesSpec and update swagger.yaml

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -359,6 +370,26 @@ instance FromJSON WalletPutData where
+instance ToJSON PostExternalTransactionData where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+instance ToJSON ByteString where
```
maybe we need to hide this stuff in `instance ToJSON PostExternalTransactionData` to avoid orphan instances coming from this :thinking:

jonathanknowles replied:
> maybe we need to hide this stuff in `instance ToJSON PostExternalTransactionData` to avoid orphan instances coming from this thinking

I agree we should definitely avoid this orphan instance.

Ideally, we should place all serialization/deserialization functionality within the `ToJSON`/`FromJSON` instances for `PostExternalTransactionData`, without needing to create instances for lower-level types that we import from other packages (such as `Text` or `ByteString`.)

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -359,6 +370,26 @@ instance FromJSON WalletPutData where
+byteStringToText :: ByteString -> Text
+byteStringToText = T.decodeUtf8 . B64.encode
+
+instance FromJSON ByteString where
```
the same remark as above

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -449,6 +452,29 @@ createTransaction w (ApiT wid) body = do
+postExternalTransaction w body = do
+    let externallySignedTx = body ^. #payload
+    tx <- liftHandler $ W.submitExternalTx w externallySignedTx
+    let outs = W.outputs @t tx -- need to check it again
```
need to check `outs`, `theAmount` and `ins` once again to make sure they are ok

paweljakubas replied:
`theAmount` is incorrect - we need inputs and changes (which are missing here). So maybe, I will use sum of outputs minus estimate fee if I can...

rvl replied:
1. If there's any significant calculation being done, I would move it to a function in Cardano.Wallet.Primitive.Model.
2. Change `theAmount` to `amt`. Or use `amount` and NamedFieldPuns. (The name shadowing warning sort of sucks if it convolutes variable names like this... maybe we could consider disabling it.)
3. To calculate amount properly, I think we need to know which output addresses belong to the sender (the change addresses).... but this we cannot know.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -356,6 +369,12 @@ data ErrSubmitTx
     | ErrSubmitTxNoSuchWallet ErrNoSuchWallet
     deriving (Show, Eq)
 
+-- | Errors occuring when submitting an externally signed transaction to the network
```
Small typo: "occuring" should be "occurring".

Also, this line is too long according to our coding standards:
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

Recommend something like:
```suggestion
-- | Errors that can occur when submitting an externally-signed transaction to
--   the network.
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -867,6 +887,15 @@ newWalletLayer tracer bp db nw tl = do
+        withExceptT ErrSubmitExternalTxNetwork $ postTx nw txWithWit
+        return tx
+
+
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -867,6 +887,15 @@ newWalletLayer tracer bp db nw tl = do
+        :: ByteString
+        -> ExceptT ErrSubmitExternalTx IO (Tx t)
+    _submitExternalTx payload = do
+        txWithWit@(tx,_) <- withExceptT ErrSubmitExternalTxDecode $ decodeExternalTx nw payload
```
```suggestion
        txWithWit@(tx,_) <- withExceptT ErrSubmitExternalTxDecode $
            decodeExternalTx nw payload
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -651,6 +677,49 @@ instance LiftHandler ErrSignTx where
             }
         ErrSignTxWithRootKey e@ErrWithRootKeyWrongPassphrase{} -> handler e
 
+
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -651,6 +677,49 @@ instance LiftHandler ErrSignTx where
+                , "transaction due to : ", pretty err
+                , "Make sure to sends a base64-encoded binary blob, in proper binary format, "
+                , "of the already serialized transaction"
+                ]
```
Suggestion for wording:
```hs
        ErrDecodeExternalTxWrongPayload err ->
            apiError err404 WrongPayload $ mconcat
                [ "I couldn't decode the payload that seems to be an "
                , "externally-signed transaction due to: ", pretty err, ". "
                , "Make sure to send a base64-encoded binary blob, in the "
                , "proper binary format of the already-serialized transaction."
                ]
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -651,6 +677,49 @@ instance LiftHandler ErrSignTx where
+            apiError err404 UnexpectedError $ mconcat
+                [ "I couldn't exert the endpoint with this backend as it is not supported"
+                , "Try this endpoint with other backend"
+                ]
```
Suggestion for wording:
```hs
        ErrDecodeExternalTxNotSupported ->
            apiError err404 UnexpectedError $ mconcat
                [ "This endpoint is not supported by the backend currently "
                , "in use. Please try a different backend."
                ]
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -651,6 +677,49 @@ instance LiftHandler ErrSignTx where
+                handler e'
+            ErrPostTxBadRequest err ->
+                apiError err500 CreatedInvalidTransaction $ mconcat
+                    [ "That's embarassing. It looks like I've created an "
```
Small typo for "embarrassing".
```suggestion
                    [ "That's embarrassing. It looks like I've created an "
```

paweljakubas replied:
corrected, also in other two places

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -13,6 +13,7 @@
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```
This is often a sign of something dangerous to come later on... :)

paweljakubas replied:
yes, this was because I added ToJSON/FromJSON instances of ByteString. Now, as everything is taking place inside `PostExternalTransactionData` instances they are not present anymore,

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -143,7 +147,9 @@ import Web.HttpApiData
 import qualified Data.Aeson.Types as Aeson
+import qualified Data.ByteString.Base64 as B64
 import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
```
Just a small point, but maybe it's worth importing this as `TE` in order to discriminate between `Data.Text` and `Data.Text.Encoding`.

paweljakubas replied:
I tried to comply with https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#hlint-is-used-for-hints-and-general-code-style
and we have there : 
```
- {name: [Data.Text, Data.Text.Encoding], as: T}
```

jonathanknowles replied:
I wasn't actually aware of that! Thanks for pointing it out.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -232,6 +242,7 @@ data ApiErrorCode
     | WalletAlreadyExists
     | NoRootKey
     | WrongEncryptionPassphrase
+    | WrongPayload
```
What's wrong about it?

paweljakubas replied:
changed to `WrongBinaryPayload` hope it better convey the message behind it

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -359,6 +370,26 @@ instance FromJSON WalletPutData where
+textToByteString :: MonadPlus m =>  Text -> m ByteString
+textToByteString x = case B64.decode (T.encodeUtf8 x) of
+    Left _ -> mzero
+    Right bs -> pure bs
```
The name of this function gives the impression of something **very general**, when in fact it's **very specific**! 

I would be in favour of _inlining_ these functions so that they're within the instance definitions, or alternatively renaming them to something that describes their purpose.

paweljakubas replied:
I inlined everything, no more ByteString instances

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -359,6 +370,26 @@ instance FromJSON WalletPutData where
+    Right bs -> pure bs
+
+byteStringToText :: ByteString -> Text
+byteStringToText = T.decodeUtf8 . B64.encode
```
See comment for `textToByteString`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -49,6 +52,10 @@ data NetworkLayer t m = NetworkLayer
+
+    , decodeExternalTx
+        :: ByteString -> ExceptT ErrDecodeExternalTx m (Tx t, [TxWitness])
+        -- ^ Decode an externally signed transaction to the chain producer
```
Just a nitpick: I would propose using hyphenation in compound adjectives such as this one. So for example: "externally-signed transaction" rather than "externally signed transaction".

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -87,6 +94,15 @@ data ErrPostTx
+
+instance Exception ErrDecodeExternalTx
+
+
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -722,6 +733,20 @@ instance Arbitrary (PostTransactionFeeData t) where
+        ]
+    arbitrary = do
+        count <- choose (0, 32)
+        BS.pack <$> replicateM count arbitrary
```
I think we should probably find a way to avoid defining an orphan `Arbitrary` instance for `ByteString` here, and instead place this logic within the `PostExternalTransactionData` `Arbitrary` instance.

rvl replied:
Or do like QuickCheck does and make a `newtype ShortByteString = ShortByteString { getShortByteString :: ByteString }`.

paweljakubas replied:
indeed! Eliminated ByteString instance by placing the logic inside  `PostExternalTransactionData` instance

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -873,6 +898,9 @@ instance ToSchema (PostTransactionData t) where
     declareNamedSchema _ = declareSchemaForDefinition "ApiPostTransactionFeeData"
 
+instance ToSchema PostExternalTransactionData where
+    declareNamedSchema _ = declareSchemaForDefinition "ApiPostExternalTransactionData"
```
```suggestion
    declareNamedSchema _ =
        declareSchemaForDefinition "ApiPostExternalTransactionData"
```

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

paweljakubas replied:
done

jonathanknowles commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Network.hs:
```diff
@@ -98,6 +99,8 @@ mkNetworkLayer httpBridge = NetworkLayer
     , networkTip = snd <$> getNetworkTip httpBridge
     , postTx = postSignedTx httpBridge
+    , decodeExternalTx = \_ ->
+            throwE ErrDecodeExternalTxNotSupported
```
Would `const $ throwE ErrDecodeExternalTxNotSupported` work here?

paweljakubas replied:
indeed! corrected

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -722,6 +733,20 @@ instance Arbitrary (PostTransactionFeeData t) where
     shrink = genericShrink
 
+instance Arbitrary PostExternalTransactionData where
+    arbitrary = genericArbitrary
```
I think we should probably avoid `genericArbitrary` if we can. It can make it hard to reason about what values are generated.

In this case, we should be able to inline your definition from `Arbitrary ByteString` below.

paweljakubas replied:
indeed. Inlined `Arbitrary ByteString` inside `PostExternalTransactionData` instance

jonathanknowles commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs:
```diff
@@ -148,6 +150,15 @@ mkNetworkLayer j = NetworkLayer
+                    throwE $ ErrDecodeExternalTxWrongPayload "wrongly constructed binary blob"
+                Right (_,_,msg) -> case msg of
+                    Transaction stx -> pure stx
+                    _ -> throwE $ ErrDecodeExternalTxWrongPayload "wrongly constructed binary blob"
```
The lines are quite a bit longer than allowed by our coding standard.

Recommend the following:

```hs
    , decodeExternalTx = \payload ->
        case runGetOrFail getMessage (BL.fromStrict payload) of
            Left _ ->
                throwE $ ErrDecodeExternalTxWrongPayload
                    "wrongly constructed binary blob"
            Right (_, _, msg) ->
                case msg of
                    Transaction stx -> pure stx
                    _ ->
                        throwE $ ErrDecodeExternalTxWrongPayload
                            "wrongly constructed binary blob"
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs:
```diff
@@ -258,6 +282,27 @@ spec = do
             runExceptT (postTx nw tx) `shouldThrow` anyException
+
+        it "decodeExternalTx works ok with properly constructed (Tx, [TxWitness]) binary blob" $
+            \(_, nw) -> do
```
```suggestion
        it "decodeExternalTx works ok with properly constructed \
            \(Tx, [TxWitness]) binary blob" $
```

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs:
```diff
@@ -258,6 +282,27 @@ spec = do
+                        `shouldReturn` Right signedTx
+
+        it "decodeExternalTx throws an exception when binary blob has non-transaction-type header \
+           \or is wrongly constructed binary blob" $
```
Suggest:
```hs
        it "decodeExternalTx throws an exception when binary blob has non-\
            \transaction-type header or is wrongly constructed binary blob" $
```

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs:
```diff
@@ -258,6 +282,27 @@ spec = do
+                            $ putSignedTx inps outs wits
+                    let encodedSignedTx = BL.toStrict $ encodeWrongly signedTx
+                    result <- runExceptT (decodeExternalTx nw encodedSignedTx)
+                    result `shouldBe` (Left $ ErrDecodeExternalTxWrongPayload "wrongly constructed binary blob")
```
Suggest:
```hs
                    result `shouldBe`
                        (Left $ ErrDecodeExternalTxWrongPayload
                            "wrongly constructed binary blob")
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs:
```diff
@@ -346,3 +391,84 @@ spec = do
+    shrink = genericShrink
+
+instance Arbitrary TxOut where
+    arbitrary = genericArbitrary
```
I would avoid using `genericArbitrary` here if you can.

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs:
```diff
@@ -346,3 +391,84 @@ spec = do
+
+-- Only generating single addresses!
+instance Arbitrary Address where
+    arbitrary = Address . prependTag 3 <$> genFixed 32
```
Where does the value `3` come from?

paweljakubas replied:
so the answer is in https://github.com/input-output-hk/cardano-wallet/blob/master/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs#L509
I am generating here single addresses that have `3` header, hence prepending with this. Moreover, I am prepending to randomly generated `32` bytes as we have https://github.com/input-output-hk/cardano-wallet/blob/master/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs#L529
Hope it answers your question

paweljakubas replied:
added to the code the following in order to avoid magic number : 
```
466         let singleAddressHeader = 3                                                                                                                                                       
467         let singleAddressLenWithoutHeader = 32
```

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -282,6 +282,11 @@ x-transactionOutputs: &transactionOutputs
       amount: *transactionAmount
 
+x-signedTransactionBlob: &signedTransactionBlob
+  description: A base64-encoded binary blob of the already signed transaction
```
```suggestion
  description: A base64-encoded binary blob of the already-signed transaction.
```

paweljakubas replied:
done

jonathanknowles reviewed (dismissed):
Hi @paweljakubas!

I've had a quick look through this PR. It mostly looks good so far. Though I still need to check that the calculations are correct. (I'll take another look tomorrow, if I can.)

Two points:

There are some orphan instances that I think we should avoid, which I've highlighted. (See comments.)

There are also some coding standards issues, which I've highlighted.

Cheers :)

rvl commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -466,10 +492,10 @@ mkApiTransaction txid ins outs meta = ApiTransaction
-    coerceTxOut (TxOut addr (Coin c)) =
-        AddressAmount (ApiT addr, Proxy @t) (Quantity $ fromIntegral c)
+
+coerceTxOut :: forall t. TxOut -> AddressAmount t
```
Better call it `addressAmountFromTxOut` or `convertTxOut`. The word "coerce" has a meaning that doesn't apply here.

paweljakubas replied:
indeed, changed

rvl commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -143,7 +147,9 @@ import Web.HttpApiData
 
 import qualified Data.Aeson as Aeson
 import qualified Data.Aeson.Types as Aeson
+import qualified Data.ByteString.Base64 as B64
```
For consistency, and for one less direct dependency, you can use `Data.ByteArray.Encoding.convertFromBase Base64`. The only problem with this function is that it's polymorphic on both ends - very slippery.

paweljakubas replied:
I promise to look at this in the second PR. now will leave it as it is as it seems to work properly

rvl commented on lib/core/src/Cardano/Wallet/Api.hs:
```diff
@@ -146,6 +148,11 @@ type ListTransactions t = "wallets"
     :> Get '[JSON] [ApiTransaction t]
 
+-- | https://input-output-hk.github.io/cardano-wallet/api/#operation/postExternalTransaction
+type PostExternalTransaction t = "externalTransactions"
```
1. As a user of the API I would prefer to post the transaction data directly with `content-type: application/octet-stream`. Rather than having to base64-encode it then stuff it into a json object. Maybe the API could support both types of request bodies if json is absolutely required.

2. And it looks like we can't give correct/useful answers for many of the fields of ApiTransaction. What about a specialised response type which just returns the txid?

paweljakubas replied:
agree 100% to the point 2. I will respond with TxId here

When it comes to `content-type: application/octet-stream` I will support it along with json in the next PR concluding this story if you are ok with this (this PR got really big). Added us Decision point to https://github.com/input-output-hk/cardano-wallet/issues/628

rvl reviewed (changes requested):
Looks pretty good, only a few suggestions from me.

paweljakubas committed:
complying with PR review

paweljakubas committed:
add integration tests for externally-signed transaction

paweljakubas committed:
polish integration tests

rvl commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -359,11 +374,31 @@ instance FromJSON WalletPutData where
     toJSON = genericToJSON defaultRecordTypeOptions
 
+instance FromJSON PostExternalTransactionData where
+    parseJSON = withObject "postExternalTransactionData" $ \o -> do
```
What about this simpler version?

```suggestion
    parseJSON = withObject "PostExternalTransactionData" $ \o -> do
        payload <- o .: "payload"
        case convertFromBase Base64 (T.encodeUtf8 payload) of
            Left e -> fail $ "Could not decode Base64 payload: " ++ show e
            Right p -> pure $ PostExternalTransactionData p
```

paweljakubas replied:
thanks for this ! It seems I eliminated base64-bytestring dependancy

rvl commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -359,11 +374,31 @@ instance FromJSON WalletPutData where
+        return $ PostExternalTransactionData p
+
+instance ToJSON PostExternalTransactionData where
+    toJSON (PostExternalTransactionData p) =
```
```suggestion
    toJSON (PostExternalTransactionData p) =
        let b64 = T.decodeUtf8 $ convertToBase Base64 p
        in object ["payload" .= Aeson.String b64 ]
```

paweljakubas replied:
done

rvl commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -203,8 +205,10 @@ spec = do
+            jsonRoundtripAndGolden $ Proxy @ApiTxId
             jsonRoundtripAndGolden $ Proxy @(PostTransactionData DummyTarget)
             jsonRoundtripAndGolden $ Proxy @(PostTransactionFeeData DummyTarget)
+            jsonRoundtripAndGolden $ Proxy @PostExternalTransactionData
```
How were these generated? Using Jormungandr?

paweljakubas replied:
there are arbitrary instances below

rvl commented on specifications/api/swagger.yaml:
```diff
@@ -1022,6 +1053,19 @@ paths:
           description: An optional filter on the address state.
       responses: *responsesListAddresses
 
+  /externalTransactions:
```
Who decided the endpoint name? This is camelCase but there is another endpoint in kebab-case. I prefer the latter for URLs.

paweljakubas replied:
so change to `external-transactions` or maybe `transactions/external` ?

paweljakubas replied:
changed to `external-transactions`

rvl commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -232,6 +246,7 @@ data ApiErrorCode
     | WalletAlreadyExists
     | NoRootKey
     | WrongEncryptionPassphrase
+    | WrongBinaryPayload
```
Can we be more precise than "Wrong"? What about MalformedTxPayload?

paweljakubas replied:
changed to `MalformedTxPayload`

rvl reviewed (changes requested):
Cool, just a few more minor change requests from me.

paweljakubas committed:
comply with second review

nix patch

hlint plus further test refactoring

apply nix patch

remove comment

adopt last fixes

rvl commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -359,11 +375,28 @@ instance FromJSON WalletPutData where
 
+instance FromJSON PostExternalTransactionData where
+    parseJSON = withObject "PostExternalTransactionData" $ \o -> do
+        (Aeson.String load) <- o .: "payload"
```
I'm pretty sure that this can decode straight to `Text` and so doesn't need the `Aeson.String` constructor.
```suggestion
        load <- o .: "payload"
```

paweljakubas replied:
yes, you are right. corrected

rvl commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -722,6 +744,17 @@ instance Arbitrary (PostTransactionFeeData t) where
+        count <- choose (0, 32)
+        bytes <- BS.pack <$> replicateM count arbitrary
+        return $ PostExternalTransactionData bytes
+    shrink (PostExternalTransactionData bytes) | BS.null bytes = []
```
I was wondering if it would be easier to just use the `instance Arbitrary [Char]` here. Like
```haskell
shrink (PostExternalTransactionData bytes) = PostExternalTransactionData . BS.pack <$> shrink (BS.unpack bytes)
```

paweljakubas replied:
make sense. fixed
